### PR TITLE
Bump llvm

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
I am starting this patch so that I can share how my changes are failing using the CI and get some help on it.

### CODE REVIEW

**Comment by nirvedhmeshram** [comment]
This builds now, however fails in lowering from linalg to refbackend. This is because the names and organization of low level passes have changed. @cathyzhyi @silvasean could you please take a look at my very last commit and give some inputs on how to change it. Currently I see an error like this
```
JIT session error: Symbols not found: [ memrefCopy ]
Expected<T> must be checked before access or destruction.
Unchecked Expected<T> contained error:
Failed to materialize symbols: { (main, { _mlir_forward, forward, _mlir_ciface_forward, refbackend_consume_func_return_mrf32, _mlir__mlir_ciface_forward, _mlir_reeturn_mrf32 }) }Aborted
```
But only on some e2e tests while other e2e tests pass.

**Comment by nirvedhmeshram** [comment]
I cant run all the tests because it just crashes with the error I mentioned above. When I try individual tests. `--filter Threshold` and `--filter TModule` passed but something like `--filter Conv2d` and `--filter HBC`  failed with the above error.
Please let me know if any specific tests you would like me to try.

**Comment by nirvedhmeshram** [comment]
Thanks for looking into this @cathyzhyi . That instruction is valid in llvm IR and should get lowered right?
https://github.com/nirvedhmeshram/llvm-project/blob/main/mlir/test/Target/LLVMIR/llvmir-intrinsics.mlir#L346

**Comment by cathyzhyi** [comment]
> https://github.com/nirvedhmeshram/llvm-project/blob/main/mlir/test/Target/LLVMIR/llvmir-intrinsics.mlir#L346

Ha, that's right. seems it should be lowered by `mlir-to-llvmir`?

**Comment by jpienaar** [comment]
That change only added a new lowering, so all the unit tests are passing? (Just in case they show a simpler repro for the failure)

**Comment by nirvedhmeshram** [comment]
I get ```
error: 'mlir-to-llvmir' does not refer to a registered pass or pass pipeline
```
Other thing to note is that in the file I shared its passed to `mlir-translate` not `mlir-opt`

**Comment by nirvedhmeshram** [comment]
@jpienaar So you mean end to end it is not guaranteed we have a flow? Since they are only showing unit tests?

**Comment by cathyzhyi** [comment]
> I get
> 
> ```
> error: 'mlir-to-llvmir' does not refer to a registered pass or pass pipeline
> ```
> 
> Other thing to note is that in the file I shared its passed to `mlir-translate` not `mlir-opt`

yea, `mlir-to-llvmir` happens in ExecutionEngine https://github.com/llvm/llvm-project/blob/84fe34a0b7fdd7bbf179981d1583693d5d5ec68b/mlir/lib/ExecutionEngine/ExecutionEngine.cpp#L245. Seems it's gonna be lowered but somehow ExecutionEngine can't resolve the intrinsic call symbol.

**Comment by jpienaar** [comment]
> @jpienaar So you mean end to end it is not guaranteed we have a flow? Since they are only showing unit tests?

I was just wondering about the unit tests as those would be smaller, so getting those all running first would highlight changes needed with a smaller input/pass surface. If those are working all, then one could tackle larger ones with hopefully most cases resorted. It just scopes the changes needed at each stage.

**Comment by nirvedhmeshram** [comment]
@jpienaar Yes unit tests are all running. Moreover the e2e tests that fail with the error I can run something like this without it giving any errors
```./build/bin/torch-mlir-opt -pass-pipeline='builtin.func(scf-bufferize),builtin.func(linalg-bufferize),func-bufferize,arith-bufferize,builtin.func(tensor-bufferize),builtin.func(finalizing-bufferize),refback-munge-calling-conventions,refback-insert-rng-globals,builtin.func(convert-linalg-to-loops),builtin.func(lower-affine),convert-scf-to-cf,builtin.func(refback-expand-ops-for-llvm),builtin.func(arith-expand),builtin.func(convert-math-to-llvm),convert-linalg-to-llvm,convert-memref-to-llvm,builtin.func(convert-arith-to-llvm),convert-std-to-llvm,reconcile-unrealized-casts' /tmp/Conv2dWithPaddingDilationStrideStaticModule.mlir```

**Comment by cathyzhyi** [comment]
> 

yea, the issue is only exposed when the IR output is processed by ExecutionEngine.  could you try if you revert https://reviews.llvm.org/D116099 would the test pass?

**Comment by nirvedhmeshram** [comment]
Actually that is not true, although it gives an IR, the IR looks like this
```module attributes {llvm.data_layout = "", torch.debug_module_name = "Conv2dWithPaddingDilationStrideStaticModule"} {
  llvm.func @abort()
  llvm.func @memrefCopy(i64, !llvm.ptr<struct<(i64, ptr<i8>)>>, !llvm.ptr<struct<(i64, ptr<i8>)>>)
more stuff
```
Is that abort function supposed to be there?

**Comment by nirvedhmeshram** [comment]
> yea, the issue is only exposed when the IR output is processed by ExecutionEngine. could you try if you revert
I can try

**Comment by cathyzhyi** [comment]
> memrefCopy

the `abort` is just a function definition. We might not call it during execution.

**Comment by nirvedhmeshram** [comment]
@cathyzhyi I reverted the patch you mentioned and it did not change the behavior

**Comment by cathyzhyi** [comment]
@nirvedhmeshram I am able to reproduce the failure with a unit test
```
torch-mlir-opt -convert-scf-to-cf -convert-arith-to-llvm -convert-memref-to-llvm -convert-std-to-llvm -reconcile-unrealized-casts $path_to_torch_mlir_src/torch-mlir/external/llvm-project/mlir/test/mlir-cpu-runner/copy.mlir | $path_to_torch_mlir_src/torch-mlir/build/bin/mlir-cpu-runner -e main -entry-point-result=void
```
JIT session error: Symbols not found: [ _mlir_ciface_print_memref_f32, memrefCopy ]
Error: Failed to materialize symbols: { (main, { main, _mlir_print_memref_f32, print_memref_f32, _mlir_main }) }

This is a unit test in mlir here https://github.com/llvm/llvm-project/blob/84fe34a0b7fdd7bbf179981d1583693d5d5ec68b/mlir/test/mlir-cpu-runner/copy.mlir#L1-L4. Notice the original test specified some path to shared libs. After removing those paths, the `memrefCopy` symbol can't be found. I think we are just missing the paths to shared libs in our e2e test when using the mlir execution engine.

**Comment by nirvedhmeshram** [comment]
That's great that we have isolated the problem. I am not sure how to provide the shared libs to e2e tests though. Dont have much experience with something like this.

**Comment by cathyzhyi** [comment]
> That's great that we have isolated the problem. I am not sure how to provide the shared libs to e2e tests though. Dont have much experience with something like this.

I don't either :D  @silvasean should be able to help us with this when he's back tmr.

**Comment by silvasean** [comment]
It looks like it is expecting this function to be linked in: https://github.com/llvm/llvm-project/blob/93d1a623cecb6f732db7900baf230a13e6ac6c6a/mlir/lib/ExecutionEngine/CRunnerUtils.cpp#L48

@MaheshRavishankar suggested that this issue is coming from linalg.copy getting lowered through memref.copy which then calls into the CRunnerUtils.cpp:memrefCopy function. The workaround is to lower linalg.copy to linalg.generic ops. Mahesh, can you link to the code in IREE that is used to convert linalg.copy to linalg.generic so we can steal it?

**Comment by MaheshRavishankar** [comment]
This should be the code. 

https://github.com/google/iree/blob/db0284d993840944cc12334ec300d1faf2db505b/iree/compiler/Codegen/Utils/Utils.cpp#L601

Do you mind if you just land this in core?

**Comment by nirvedhmeshram** [comment]
Thanks @silvasean and @MaheshRavishankar 
I can land this in core. Should I first just try to fix it for torch-mlir and try to get this PR merged in before adding it into the core?

**Comment by nirvedhmeshram** [comment]
Also, just went through one of the failing IR's in linalg and it does not have `linalg.copy`, could there be other ops that end up making `memref.copy` ?

**Comment by nirvedhmeshram** [comment]
For example
 ```
%28 = arith.cmpi sge, %27, %19 : index
%29 = arith.select %28, %27, %19 : index
%30 = arith.subi %29, %19 : index
 %31 = tensor.extract_slice %arg1[%19] [%30] [1] : tensor<?xi32> to tensor<?xi32>
```
After `builtin.func(linalg-bufferize)'` pass becomes
```
%30 = arith.cmpi sge, %29, %21 : index
%31 = arith.select %30, %29, %21 : index
%32 = arith.subi %31, %21 : index   
 %33 = memref.alloc(%32) : memref<?xi32>
 %34 = memref.subview %1[%21] [%32] [1] : memref<?xi32> to memref<?xi32, #map2>
 memref.copy %34, %33 : memref<?xi32, #map2> to memref<?xi32>
 ```

**Comment by MaheshRavishankar** [comment]
> For example
> 
> ```
> %28 = arith.cmpi sge, %27, %19 : index
> %29 = arith.select %28, %27, %19 : index
> %30 = arith.subi %29, %19 : index
> %31 = tensor.extract_slice %arg1[%19] [%30] [1] : tensor<?xi32> to tensor<?xi32>
> ```
> 
> After `builtin.func(linalg-bufferize)` pass becomes
> 
> ```
> %30 = arith.cmpi sge, %29, %21 : index
> %31 = arith.select %30, %29, %21 : index
> %32 = arith.subi %31, %21 : index   
> %33 = memref.alloc(%32) : memref<?xi32>
> %34 = memref.subview %1[%21] [%32] [1] : memref<?xi32> to memref<?xi32, #map2>
> memref.copy %34, %33 : memref<?xi32, #map2> to memref<?xi32>
> ```

Yeah, bufferization by default adds `memref.copy`. I dont know what `linalg-bufferize` calls, but if this is the upstream bufferization you need to update the callbacks to use the right allocation/deallocation/copy functions (like here https://github.com/google/iree/blob/db0284d993840944cc12334ec300d1faf2db505b/iree/compiler/Codegen/LLVMCPU/Passes.cpp#L284)

**Comment by silvasean** [comment]
> For example
> 
> ```
> %28 = arith.cmpi sge, %27, %19 : index
> %29 = arith.select %28, %27, %19 : index
> %30 = arith.subi %29, %19 : index
> %31 = tensor.extract_slice %arg1[%19] [%30] [1] : tensor<?xi32> to tensor<?xi32>
> ```
> 
> After `builtin.func(linalg-bufferize)` pass becomes
> 
> ```
> %30 = arith.cmpi sge, %29, %21 : index
> %31 = arith.select %30, %29, %21 : index
> %32 = arith.subi %31, %21 : index   
> %33 = memref.alloc(%32) : memref<?xi32>
> %34 = memref.subview %1[%21] [%32] [1] : memref<?xi32> to memref<?xi32, #map2>
> memref.copy %34, %33 : memref<?xi32, #map2> to memref<?xi32>
> ```

Ugh, yeah, that's tricky. Could you try to write a pass to convert memref.copy to linalg.copy? You would insert the pass [here](https://github.com/llvm/torch-mlir/blob/869daf3c227d0e09604b93b56fbf45d30740af81/python/torch_mlir_e2e_test/linalg_on_tensors_backends/refbackend.py#L181). linalg-to-loops should take care of lowering it then. I would call the pass `refback-munge-memref-copy`.

**Comment by nirvedhmeshram** [comment]
Sounds good, yes I can do that.

File: lib/Conversion/TorchToLinalg/TorchToLinalg.cpp [hunk 1]
```diff
@@ -30,6 +32,7 @@ using namespace mlir::torch;
 using namespace mlir::torch::Torch;
 using namespace mlir::torch::TorchConversion;
 using namespace mlir::torch::torch_upstream; // For ScalarType and type
+using namespace mlir::cf;
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/TorchToLinalg.cpp:35]: I would prefer to put the dialect prefix before the operation.

File: lib/RefBackend/RefBackend.cpp [hunk 2]
```diff
@@ -354,3 +357,65 @@ std::unique_ptr<OperationPass<FuncOp>>
...
+
+class MungeMemrefCopy : public MungeMemrefCopyBase<MungeMemrefCopy> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect>();
+  }
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    RewritePatternSet patterns(&getContext());
+    // ConversionTarget target(*context);
```
Comment (cathyzhyi) [inline lib/RefBackend/RefBackend.cpp:408]: nit: remove this line.

**Review by cathyzhyi — APPROVED** [review]
Great work!! Thanks a lot for tackle this complicated llvm bump. 
Just 2 nits.

**Pull request merged**